File: Έγγραφα SEM4105/Λύσεις ασκήσεων/10. Δέκατο εργαστήριο/04/taylor.py
import math

def difeq(x,y):
    dy=list((0,0,0,0,0))

    dy[0] = math.cos(x) - math.sin(y) + x*x
    dy[1] = 2.0 * x - math.sin(x) - math.cos(y) * dy[0]    
    dy[2] = 2.0 - math.cos(x) -  math.cos(y) * dy[1] + math.sin(y) * dy[0]**2
    
    dy[3] =  math.sin(x) + 3.0 *  math.sin(y) * dy[0] * dy[1] + math.cos(y) * (dy[0]**3 - dy[2])
	
    dy[4] =  math.cos(x) + math.cos(y) * (6.0 * dy[0]**2 * dy[1] - dy[3]) + (3.0 * dy[1]**2 + dy[0] * (4.0 * dy[2] - dy[0]**3)) * math.sin(y)

    return dy


def taylorstep(x0,y0,x1):
    dy = difeq(x0,y0)
    y1 =y0
    h = x1-x0
    term = h

    for i in range(len(dy)):
        y1 += term * dy[i]
        term *= h / (i+2)

    return y1
    



def taylor(a,b,h,ya):
    xold = a
    yold = ya

    while xold < b:
        xnew = xold + h
        if xnew > b:
            xnew = b
	
        ynew = taylorstep(xold, yold, xnew)

        xold = xnew
        yold = ynew
    
    return ynew



def main():
    xa = -1.0
    xb = 1.0
    h = 0.01
    ya = 3.0
    yb = taylor(xa, xb, h, ya)
    
    print(xb, yb)




main()

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/10. Δέκατο εργαστήριο/04/taylor.cpp
#include <cmath>
#include <vector>
#include <iostream>


void difeq(double x, double y, std::vector<double> & dy)
{    
    using std::sin;
    using std::cos;
    
    dy[0] = cos(x)-sin(y) + x*x;
    
    dy[1] = 2.0 * x - sin(x) - cos(y) * dy[0];
    
    dy[2] = 2.0 - cos(x) - cos(y) * dy[1] + sin(y) * dy[0]*dy[0];
    
    dy[3] = sin(x) + 3.0 * sin(y) * dy[0] * dy[1] +
	cos(y) * (std::pow(dy[0],3) - dy[2]);	

    dy[4] = cos(x) + cos(y) * (6.0 * dy[0] * dy[0] * dy[1] - dy[3]) 
	+ (3.0 * dy[1] * dy[1] + dy[0] * (4.0 * dy[2] - std::pow(dy[0],3)))
	* sin(y);
}

void taylorstep(double x0, double y0, double x1, double & y1)
{
    std::vector<double> dy(5);
    
    difeq(x0, y0, dy);
    
    y1 = y0;
    double h = x1-x0;
    double term = h;
    
    for (std::size_t i = 0; i < dy.size(); ++i) {
	y1 += dy[i] * term;
	term *= h / (i+2);
    }

}


void taylor(double a, double b, double h, double ya, double & yb)
{
    double xold = a;
    double yold = ya;
    double ynew;

    while (xold < b) {
	double xnew = xold + h;
	if (xnew > b) {
	    xnew = b;
	}
	
	taylorstep(xold, yold, xnew, ynew);

	xold = xnew;
	yold = ynew;
    }

    yb = ynew;
}


int main()
{
    double xa = -1.0;    
    double xb = 1.0;
    double h = 0.01;
    double ya = 3.0;
    double yb;
    
    taylor(xa, xb, h, ya, yb);

    std::cout.precision(15);    
    std::cout << xb << ' ' << yb << '\n'; 
}
